doc/design/request_response_communication.md

```diff
@@ -0,0 +1,118 @@
+| Client            | an entity invoking RPCs                                  |
+| Server            | an entity processing RPCs                                |
+
+## Design
+
+### Considerations
+
+Client and server shall be able to be used in combination with the `WaitSet` and `Listener`
+to provide a means of event based communication instead of polling.
+
+Like with publish-subscribe, the `ServiceDescription` will be used to connect clients with a corresponding server.
+
+In order to support asynchronous requests, a sequence ID should be part of each request which will be copied to the corresponding response.
+
+### Solution
+
+This is an overview of the untyped `Client` and `Server` classes.
+
+![simple class diagram](diagrams/request_response/overview_class.svg)
+
+The `Client` and `Server` are reusing the `ChunkSender` and `ChunkReceiver` building blocks. The `Client` uses a `ChunkSender` to send requests and a `ChunkReceiver` to get the responses while the `Server` uses a `ChunkReceiver` to get the requests and a `ChunkSender` to send the responses.
+
+#### Typed API
+
+![typed API](diagrams/request_response/typed_api.svg)
+
+Since the `Response` is tied to a specific `Request`, the `loan` method takes a `Request` to populate the `Response` with the correct settings.
+Alternatively, a `Request` could have a `createResponse` method which returns a `Response` with the correct settings.
+
+#### Untyped API
+
+![untyped API](diagrams/request_response/untyped_api.svg)
+
+Similar to the the typed API, `loan` takes a pointer to a `RequestHeader` to populate the `ResponseHeader` with the correct settings.
+
+#### Client Port
+
+![client port](diagrams/request_response/client_port.svg)
+
+The `ClientPortData` is located in the shared memory and contain only the data but no methods to access them.
+`ClientPortUser` is the class providing the methods for the user access and `ClientPortRouDi` provides the
+interface RouDi needs to connect the client to the server and to cleanup the port resources.
+
+#### Server Port
+
+![server port](diagrams/request_response/server_port.svg)
+
+Similar to the Client Port, the Server Port has `ServerPortData` which is located in the shared memory and contain only the data but no methods to access them.
+`ServerPortUser` is the class providing the methods for the user access and `ServerPortRouDi` provides the
+interface RouDi needs to connect the client to the server once the server offers its service and to cleanup the port resources.
+
+It must be ensured that only one server with a given `ServiceDescription` can run at a time.
+
+#### Request/Response Header
+
+![rpc header](diagrams/request_response/request_response_header.svg)
+
+Since request and response need to encode different meta-information, we also need different header for the messages.
+The common data is aggregated in `RpcBaseHeader` which contains a `UniquePortId` to the port owning the `ClientChunkQueueData_t` and a sequence ID.
```

Comment:
Note that by coupling request and response as in my future based proposal (comment above), the sequence ID would probably be unnecessary (the client essentially knows where to find the response, the server knows where to write it). If it issued 3 requests and gets notified of one response, it has to look however which request is ready (in a naive implementation).

But yes, for out of order responses a sequence ID is necessary in the way described here.